### ISSUE TITLE
[Question] about gradient_accumulate_steps

### ISSUE THREAD

**Issue by Jayce-Ping** [yifan123/flow_grpo#155]
Hi, thanks for your excellent work!

I've been thinking about this question for a long time. Why do we choose 2-step updates for each epoch? Intuitively, 1-step should provide the largest equivalent batch_size and should work best. However, I remember that you mentioned somewhere in the issue that you found 2-step works best.

I think, during the first step, the model has not been updated after sampling, which should have the exactly same weights as before. During this period, the log_prob of `old` model and `new` model should be exact same. The difference in actual training is from `model.train()`/`model.eval()` difference and maybe some numerical precision. So, in step 1, the model weight is updated to be mildly different from sampling weights. Then, the second step really updates the model.

A comparison is [DiffusionNFT](https://github.com/NVlabs/DiffusionNFT/tree/main/scripts), which decouples the sampling and training models and set `gradient_accumulate_steps_per_epoch=1` as default setting. Here, the `old` and `new` are inherently different and makes real updates.

I used two `lora` modules (`old` and `new`) to represent old policy and new policy, and skip the `log_prob` calculation during sampling. Before the training loop, there are exactly the same. During the training loop, `old` lora is frozen and `new` will be updated. I observed that the `ratio` will be excat `1` for every odd step, since they are the same before the first update.

For this potential issue, a possible solution is to use the `ema` version of the `new` policy as the `old` policy, inspired by [DiffusionNFT](https://github.com/NVlabs/DiffusionNFT/tree/main/scripts). I'm still working on some experiments to see if it helps training.

Am'I understanding something wrong here? What do you think?

**Comment by yifan123** [comment]
I have tried a specific setup: Assume the batch size (bs) is set to b, collect data for n iterations, then concatenate this data into a single batch with a size of b × n, and recalculate the log probability afterward. For training, this large batch is also used to run two training iterations. This approach aligns with the current practices for Large Language Models (LLMs) and has shown good performance.
The claim that "1-step training fails to converge" was my initial finding, but at that time, other hyperparameters might not have been properly tuned. Someone has informed me that their self-implemented on-policy training for FlowGRPO runs stably. You can try pure on-policy training on the latest version of the repository.

**Comment by MengHao666** [comment]
> Hi, thanks for your excellent work!
> 
> I've been thinking about this question for a long time. Why do we choose 2-step updates for each epoch? Intuitively, 1-step should provide the largest equivalent batch_size and should work best. However, I remember that you mentioned somewhere in the issue that you found 2-step works best.
> 
> I think, during the first step, the model has not been updated after sampling, which should have the exactly same weights as before. During this period, the log_prob of `old` model and `new` model should be exact same. The difference in actual training is from `model.train()`/`model.eval()` difference and maybe some numerical precision. So, in step 1, the model weight is updated to be mildly different from sampling weights. Then, the second step really updates the model.
> 
> A comparison is [DiffusionNFT](https://github.com/NVlabs/DiffusionNFT/tree/main/scripts), which decouples the sampling and training models and set `gradient_accumulate_steps_per_epoch=1` as default setting. Here, the `old` and `new` are inherently different and makes real updates.
> 
> I used two `lora` modules (`old` and `new`) to represent old policy and new policy, and skip the `log_prob` calculation during sampling. Before the training loop, there are exactly the same. During the training loop, `old` lora is frozen and `new` will be updated. I observed that the `ratio` will be excat `1` for every odd step, since they are the same before the first update.
> 
> For this potential issue, a possible solution is to use the `ema` version of the `new` policy as the `old` policy, inspired by [DiffusionNFT](https://github.com/NVlabs/DiffusionNFT/tree/main/scripts). I'm still working on some experiments to see if it helps training.
> 
> Am'I understanding something wrong here? What do you think?

Could u provide ur code? I found all my ratio value is 1

**Comment by Jayce-Ping** [comment]
> > Hi, thanks for your excellent work!
> > I've been thinking about this question for a long time. Why do we choose 2-step updates for each epoch? Intuitively, 1-step should provide the largest equivalent batch_size and should work best. However, I remember that you mentioned somewhere in the issue that you found 2-step works best.
> > I think, during the first step, the model has not been updated after sampling, which should have the exactly same weights as before. During this period, the log_prob of `old` model and `new` model should be exact same. The difference in actual training is from `model.train()`/`model.eval()` difference and maybe some numerical precision. So, in step 1, the model weight is updated to be mildly different from sampling weights. Then, the second step really updates the model.
> > A comparison is [DiffusionNFT](https://github.com/NVlabs/DiffusionNFT/tree/main/scripts), which decouples the sampling and training models and set `gradient_accumulate_steps_per_epoch=1` as default setting. Here, the `old` and `new` are inherently different and makes real updates.
> > I used two `lora` modules (`old` and `new`) to represent old policy and new policy, and skip the `log_prob` calculation during sampling. Before the training loop, there are exactly the same. During the training loop, `old` lora is frozen and `new` will be updated. I observed that the `ratio` will be excat `1` for every odd step, since they are the same before the first update.
> > For this potential issue, a possible solution is to use the `ema` version of the `new` policy as the `old` policy, inspired by [DiffusionNFT](https://github.com/NVlabs/DiffusionNFT/tree/main/scripts). I'm still working on some experiments to see if it helps training.
> > Am'I understanding something wrong here? What do you think?
> 
> Could u provide ur code? I found all my ratio value is 1

@MengHao666  If you set `gradient_accumulate_steps_per_epoch =1`, the ratio may always be 1. You can find my code at [my repo](https://github.com/X-GenGroup/PaCo-RL/tree/main/PaCo-GRPO). I used two lora adapter to be compatible with `Diffusion-NFT` loss.
